### PR TITLE
RavenDB-20207: add ability to set CV in subscription, fix possible stale

### PR DESCRIPTION
**Pull request by garayx**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20207

### Additional description

Add ability to set the CV to LastDocument (and ability to set CV by admin (only serverside))
Fix possible stale of subscription, wasn't setting result to the tcs of subscription batch

### Type of change

- Regression bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs [hunk 1]
```diff
@@ -137,6 +137,10 @@ public override unsafe void Execute(ClusterOperationContext context, Table items
                     shouldUpdateChangeVector = false;
                 }
 
+                if (string.IsNullOrEmpty(ShardName) == false)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs]: debug code?

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
@garayx Please check failing tests

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs [hunk 2]
```diff
@@ -32,6 +33,46 @@ public ShardedSubscriptionWorker(SubscriptionWorkerOptions options, string dbNam
             };
         }
 
+        internal override async Task<BatchFromServer> ReadSingleSubscriptionBatchFromServerAsync(JsonContextPool contextPool, Stream tcpStream, JsonOperationContext.MemoryBuffer buffer, ShardedSubscriptionBatch batch)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs]: this is under the `PrepareBatchAsync`, no?

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionBatch.cs [hunk 3]
```diff
@@ -52,11 +52,19 @@ internal override async ValueTask InitializeAsync(BatchFromServer batch)
         Context = batch.Context;
         batch.ReturnContext = null; // move the release responsibility to the OrchestratedSubscriptionProcessor
         batch.Context = null;
-  
-        await InitializeDocumentIncludesAsync(batch);
-        await base.InitializeAsync(batch);
 
-        LastSentChangeVectorInBatch = null;
+        try
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionBatch.cs]: isn't it part if `PrepareBatchAsync`?

File: src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs [hunk 4]
```diff
@@ -34,6 +37,14 @@ public override void HandleDatabaseRecordChange(DatabaseRecord databaseRecord)
                 }
 
                 var taskState = JsonDeserializationClient.SubscriptionState(taskStateRaw);
+
+                if (SubscriptionChangeVectorHasChanges(state, taskState))
+                {
+                    DropSubscriptionConnections(id, new SubscriptionClosedException(
+                        $"The subscription '{subscriptionName}' change vector was modified on shard '{_shardName}', connection must be restarted",
+                        canReconnect: true));
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs:46]: return ?

File: src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs [hunk 5]
```diff
@@ -44,6 +55,34 @@ public override void HandleDatabaseRecordChange(DatabaseRecord databaseRecord)
         }
     }
 
+    protected override bool SubscriptionChangeVectorHasChanges(SubscriptionConnectionsState state, SubscriptionState taskStatus)
+    {
+        if (taskStatus.LastClientConnectionTime != null)
+            return false;
+
+        if (taskStatus.ShardingState == null)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs]: is it even valid here?

anyway let's have a cleaner code here
```
if(taskStatus.ShardingState == null)
  return base()
  
...
```

File: src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs [hunk 6]
```diff
@@ -24,22 +25,51 @@ private PutShardedSubscriptionCommand()
 
     protected override void HandleChangeVectorOnUpdate(ClusterOperationContext context, SubscriptionState existingSubscriptionState, long subscriptionId)
     {
+        var existingShardsCVs = existingSubscriptionState.ShardingState.ChangeVectorForNextBatchStartingPointPerShard;
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs:29]: can this be `null`?

File: src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs [hunk 7]
```diff
@@ -109,6 +114,19 @@ private bool IsLegacyCommand()
                    BatchId == ISubscriptionConnection.NonExistentBatch; // from noop ack
         }
 
+        private void CheckConcurrencyForBatchCv(SubscriptionState state, string subscriptionName)
+        {
+            if (state.ShardingState.ChangeVectorForNextBatchStartingPointPerShard.TryGetValue(ShardName, out string cvInStorage))
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:119]: we don't have this check for normal database, so why for sharded?

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/ServerWide/Commands/Subscriptions/RecordBatchSubscriptionDocumentsCommand.cs]: removed

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs (hunk 2 above)
Comment (garayx) [reply src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs]: yup

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionBatch.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionBatch.cs]: done

File: src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs (hunk 4 above)
Comment (garayx) [reply src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs:46]: yes,       continue;

File: src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs (hunk 5 above)
Comment (garayx) [reply src/Raven.Server/Documents/Subscriptions/Sharding/ShardSubscriptionStorage.cs]: should not happen, I will add a debug assert there

File: src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs (hunk 6 above)
Comment (garayx) [reply src/Raven.Server/ServerWide/Commands/Sharding/PutShardedSubscriptionCommand.cs:29]: no for sharded subscription existingSubscriptionState.ShardingState is always not null.

File: src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs (hunk 7 above)
Comment (garayx) [reply src/Raven.Server/ServerWide/Commands/Subscriptions/AcknowledgeSubscriptionBatchCommand.cs:119]: this is here to drop the connection faster (even on noop ack) if the CV is changed (same logic as `ResponsibleNode` & delete database few lines above),

I guess it was removed for the regular database when concurrent subscriptions were added, thats why you dont have this check for regular db

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs [hunk 8]
```diff
@@ -603,29 +556,77 @@ private async Task ProcessSubscriptionInternalAsync(JsonContextPool contextPool,
...
+            }
+        }
+
+        protected virtual void HandleSubscriberError(Exception ex)
+        {
+            if (_options.IgnoreSubscriberErrors == false)
+                throw new SubscriberErrorException($"Subscriber threw an exception in subscription '{_options.SubscriptionName}'", ex);
+        }
+
+        internal virtual async Task<BatchFromServer> PrepareBatchAsync(JsonContextPool contextPool, Stream tcpStreamCopy, JsonOperationContext.MemoryBuffer buffer, TBatch batch, Task notifiedSubscriber)
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs:577]: protected?

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs [hunk 9]
```diff
@@ -746,19 +747,35 @@ internal async Task<SubscriptionConnectionServerMessage> ReadNextObjectAsync(Jso
             }
         }
 
-        internal static async Task SendAckAsync(string lastReceivedChangeVector, Stream stream, JsonOperationContext context, CancellationToken token)
+        internal virtual async Task SendAckAsync(TBatch batch, Stream stream, JsonOperationContext context, CancellationToken token)
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs]: protected?

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs [hunk 10]
```diff
@@ -746,19 +747,35 @@ internal async Task<SubscriptionConnectionServerMessage> ReadNextObjectAsync(Jso
...
+            }
+            catch (ObjectDisposedException)
+            {
+                //if this happens, this means we are disposing, so don't care..
+                //(this piece of code happens asynchronously to external using(tcpStream) statement)
+            }
+        }
 
-            using (var messageJson = DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(message, context))
+        protected async Task SendAckInternalAsync(TBatch batch, Stream stream, JsonOperationContext context, CancellationToken token)
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs:763]: why do we have `SendAckInternalAsync` & `SendAckAsync`? looks identical, no?

File: src/Raven.Server/Documents/Sharding/Subscriptions/SubscriptionConnectionsStateOrchestrator.cs [hunk 11]
```diff
@@ -120,6 +121,19 @@ public override void DropSubscription(SubscriptionException e)
         _databaseContext.SubscriptionsStorage.DropSubscriptionConnections(SubscriptionId, e);
     }
 
+    public override async Task HandleConnectionException(OrchestratedSubscriptionConnection connection, Exception e)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Subscriptions/SubscriptionConnectionsStateOrchestrator.cs]: HandleConnectionExceptionAsync

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs (hunk 8 above)
Comment (garayx) [reply src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs:577]: no we use `BatchFromServer` which is internal

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs (hunk 9 above)
Comment (garayx) [reply src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs]: done

File: src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs (hunk 10 above)
Comment (garayx) [reply src/Raven.Client/Documents/Subscriptions/AbstractSubscriptionWorker.cs:763]: no, we have different exception handling in `SendAckAsync`

File: src/Raven.Server/Documents/Sharding/Subscriptions/SubscriptionConnectionsStateOrchestrator.cs (hunk 11 above)
Comment (garayx) [reply src/Raven.Server/Documents/Sharding/Subscriptions/SubscriptionConnectionsStateOrchestrator.cs]: done

**Pull request merged**